### PR TITLE
add barabasi albert model

### PR DESCRIPTION
**Pull request by afternone**
```
using LightGraphs
using StatsBase
using PyPlot

julia> @time g = barabasi_albert(100_000, 3)
  0.376023 seconds (2.88 M allocations: 105.213 MB, 9.96% gc time)
{100000, 299991} undirected graph

julia> c = countmap(degree(g))
julia> loglog(collect(keys(c)), collect(values(c)), "o")
julia> xlabel("degree")
julia> ylabel("count")
```
BA network

### CODE REVIEW

**Comment by CarloLucibello** [comment]
Hi, thanks for this nice additions. Two things: Can you add the function signature to the docstring? Also can you  use a non-global rng? take a look to function `random_regular_graph` as an example

File: src/randgraphs.jl [hunk 1]
```diff
@@ -148,6 +148,48 @@ function _try_creation(n::Int, k::Vector{Int}, rng::AbstractRNG)
     return edges
 end
 
+"""Creates a [Barabási–Albert model](https://en.wikipedia.org/wiki/Barab%C3%A1si%E2%80%93Albert_model) 
+random graph with `n` nodes is grown by attaching new nodes each with `m` edges that 
+are preferentially attached to existing nodes with high degree. Undirected graphs are
+created by default; use `is_directed=true` to override.
+"""
+function barabasi_albert(n::Integer, m::Integer; is_directed=false)
```
Comment (CarloLucibello) [inline src/randgraphs.jl]: can you rename `m` to `k` or `z`? `m` is misleading since it is often used for the total nuber of edges in the graph

**Comment by CarloLucibello** [comment]
I see something that can be improved here: 
the `repeated_nodes` vector grows to size O(n^2), we can do better. 
To sample form discrete distributions you can use the functions `sample(weightvec)` from StatsBase.
You will have to substitute this lines
```julia
push!(repeated_nodes, source)
push!(repeated_nodes, target)
```
with 
```julia
weightvec[source] += 1 
weightvec[target] += 1
```

**Comment by afternone** [comment]
@CarloLucibello The `weightvec` way cost about 26s for the above example network.

**Comment by afternone** [comment]
How to pass the `rng` to sample?

**Comment by CarloLucibello** [comment]
>@CarloLucibello The weightvec way cost about 26s for the above example network.

omg, so the implementation in StatsBase is highly inefficient. Ok so live it the way it is

 

>How to pass the rng to sample?

There is a sample function in src/utils.jl that accepts a `rng`, maybe you can try that. If this option doesn't work just use the global rng, add an optional seed and the line 
`seed > 0 && srand(seed)`

**Comment by CarloLucibello** [comment]
good to go!

**Comment by jpfairbanks** [comment]
Can you make a minimum working example showing that SampleBase's sampling is slow. We should file that as an issue to their repo.

**Comment by sbromberger** [comment]
@jpfairbanks since the MWE doesn't hold up this PR, I'm gonna go ahead and merge.

**Pull request merged**